lua/nvim-tree/renderer/decorator/size.lua

```diff
@@ -0,0 +1,134 @@
+local HL_POSITION = require("nvim-tree.enum").HL_POSITION
+local ICON_PLACEMENT = require("nvim-tree.enum").ICON_PLACEMENT
+local view = require "nvim-tree.view"
+
+local Decorator = require "nvim-tree.renderer.decorator"
+
+---@class DecoratorSize: Decorator
+---@field width_cutoff integer
+---@field noshow_folder_size_glyph string
+---@field column_width integer
+---@field show_folder_size boolean
+---@field units string[]
+---@field format_unit function
+---@field format_size function
+local DecoratorSize = Decorator:new()
+
+---@param opts table
+---@return DecoratorSize
+function DecoratorSize:new(opts)
+  local o = Decorator.new(self, {
+    enabled = opts.renderer.size.enable,
+    hl_pos = HL_POSITION.none,
+    icon_placement = ICON_PLACEMENT.right_align,
+  })
+  ---@cast o DecoratorSize
+
+  if not o.enabled then
+    return o
+  end
+
+  o.width_cutoff = opts.renderer.size.width_cutoff
+  o.noshow_folder_size_glyph = opts.renderer.size.noshow_folder_size_glyph or ""
+  o.column_width = opts.renderer.size.column_width
+  o.show_folder_size = opts.renderer.size.show_folder_size
+  o.units = { "B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB" }
+
+  -- Guaranteed to be a function from previous setup
+  o.format_unit = opts.renderer.size.format_unit
+  o.format_size = function(size)
+    local formatted = string
+      .format("%.2f", size)
+      -- Remove trailing zeros after the decimal point
+      :gsub("(%..-)0+$", "%1")
+      -- Remove the trailing decimal point if it's a whole number
+      :gsub("%.$", "")
+    return formatted
+  end
+
+  return o
+end
+
+--- Convert a size to a human-readable format (e.g., KB, MB, GB) with fixed width
+---@private
+---@param size number size in bytes
+---@return string
+--- edit: Since then I've moved a lot of ifs around getting down to only three, but I'll keep the comment bellow just in case.
+---NOTE: This function still try it's best to minified the string
+--- generated, but this implies that we have some branches
+--- to determined how much bytes can we shave from the string to
+--- comply to `self.column_width`.
+--- Since we know `self.column_width` doesn't change, a better way could be
+--- decide a version of 'human_readable_size' a priori, based
+--- on `self.column_width` *once* at this object's construction.
+--- Basically, instead of this method, we would "baking" all ifs first
+--- to decide which function to bind to possible field `self.human_readable_size`.
+function DecoratorSize:human_readable_size(size)
+  -- Check for nan, negative, etc.
+  if type(size) ~= "number" or size ~= size or size < 0 then
+    return ""
+  end
+  local index = 1
+
+  -- We check for index here because for exaple, let's say
+  -- on a given iteration you have :
+  -- 1024 YB, then index is equal to #units, normally we would devide again,
+  -- but we don't have more units, so keep as is.
+  while size >= 1024 and index < #self.units do
```

Comment:
Could we do something really dumb like:
* convert size to a string
* calculate string length
* div 3 to get the index